text/0026-sassafras-consensus.md

```diff
@@ -0,0 +1,1079 @@
+indicating the validity of the signature (`True` if it's valid). It's important
+to note that this function does not require the signer's public key.
+
+```rust
+    BandersnatchRingVerifierKey ::= OCTET_STRING;
+
+    fn ring_vrf_verify(
+        verifier: BandersnatchRingVerifierKey,
+        signature: RingVrfSignature,
+    ) -> BOOLEAN;
+```
+
+In this document, the types `BandersnatchRingProverKey`,
+`BandersnatchRingVerifierKey`, and `RingSignature` are intentionally left
+undefined. Their definitions can be found in the `bandersnatch_vrfs` reference
+implementation.
+
+
+## 6. Sassafras Protocol
+
+### 6.1. Epoch's First Block
+
+The first block produced for epoch `N` is required to include the descriptor for
+the next epoch `N+1`.
+
+The descriptor for next epoch is `NextEpochDescriptor`.
+   
+```rust
+    AuthorityId ::= BandersnatchPublicKey;
+
+    Randomness ::= OCTET_STRING(SIZE(32));
+
+    NextEpochDescriptor ::= SEQUENCE {
+        randomness: Randomness,
+        authorities: SEQUENCE_OF AuthorityId,
+        configuration: ProtocolConfiguration OPTIONAL
+    }
+```
+
+- `randomness`: randomness value.
+- `authorities`: list of authorities.
+- `configuration`: optional protocol configuration.
+
+The `NextEpochDescriptor` must be `SCALE` encoded and embedded in the block
+header digest log.
+
+The identifier for the digest element is `BYTES("SASS")`.
+
+**Security Consideration**: Instances of `NextEpochDescriptor` are generated
+through on-chain code whenever a block is identified as the first of an epoch.
+Consequently, every node executing the block should verify that the descriptor
+locally generated during block execution matches the one produced by the block
+author, which is found in the digest data before block import.
+
+#### 6.1.1. Epoch Randomness
+
+Each block ships with some entropy source in the form of bandersnatch
+`VrfOutput`. Per block randomness is accumulated in the protocol's on-chain
+state **after** block import.
+
+The exact procedure to accumulate per-block randomness is described in detail
+later, in the [Randomness Accumulator] paragraph.
+
+Next epoch `randomness` is computed as:
+
+```rust
+    BLAKE2(32, CONCAT(accumulator, curr_epoch_randomness, next_epoch_index));
+```
+
+<TODO>
+Do we really need need to add `curr_epoch_randomness` to the hash?
+
+Looks like it doesn't add any extra entropy to the `next_epoch_randomness` as we
+are already using `accumulator`
+
+    curr_epoch_randomness = H(prev_accumulator ++ ... ++ curr_epoch_index)
+    next_epoch_randomness = H(curr_accumulator ++ curr_epoch_randomness ++ next_epoch_index)
+
+    With:
+    - curr_accumulator depending on prev_accumulator
+    - next_epoch_index depending on curr_epoch_index 
+
+    So the only extra thing curr_epoch_randomness brings to the table is `...`.
+    But if we recursively unroll this we end up `...` being the genesis randomness
+    which is always 0.
+</TODO>
```

Comment:
@burdges I need your opinion about this snip I've marked with `TODO`.

In short here I ask (in the context of computing the randomness of next epoch) what is the point of adding current epoch randomness when this is a function of the accumulator (which we already use in the hash).

Well at rigor, current epoch randomness is a function of the accumulator one epoch ago, but current accumulator is a function of it as well.

I hope my point 🫨

Reply:
It's fine if `accumulator` just accumulates forever.  We need `curr_epoch_randomness` to be copied from `accumulator` far enough in the past that we have consensus on its current value even if grandpa dies.  So:

Every block does `accumulator_after_block = H(accumulator_before_block, randomness)` like you describe below.

An epoch end does
```
curr_epoch_randomness = next_epoch_randomness;
next_epoch_randomness = H(accumulator_before_block,next_epoch_index); 
```
Including `next_epoch_index` in this hash means we can leave it out of the `vrf_input` hashes.

Does the after vs before matter here?  We obtain finality by waiting for a whole epoch, which becomes a sloppy notion, so they're not overly critical.  If otoh someone can rewind the chain via disputes or whatever, then corner cases matter, so maybe it easier to just be precises here that `next_epoch_randomness` gets get from the `accumulator` coming out of the final block of the previous epoch.  It's likely easier to code this way.